### PR TITLE
Implement Optional Private Keys

### PR DESCRIPTION
**Pull request by ferrine**
Address #160 

- [x] creation form update
<details>
image
</details>

- [x] validate public key
<details>
image
</details>

- [x] validate preshared key
<details>
image
</details>

- [x] check QR code creation, disable QR if no private key
<details>
image
</details>

- [x] check public key duplication
<details>
image
</details>

- [x] make the additional fields hidden by default
<details>
image
</details>

### CODE REVIEW

**Comment by ferrine** [comment]
Ready for review

File: templates/base.html [hunk 1]
```diff
@@ -184,6 +184,20 @@ <h4 class="modal-title">New Wireguard Client</h4>
...
                             </div>
+                            <div class="form-group">
+                                <label for="client_public_key" class="control-label">
+                                    Public Key
+                                </label>
+                                <input type="text" class="form-control" id="client_public_key" 
+                                    name="client_public_key" placeholder="Autogenerated">
+                            </div>
+                            <div class="form-group">
+                                <label for="client_public_key" class="control-label">
```
Comment (SomeoneSerge) [inline templates/base.html:195]: `client_preshared_key`?

**Review by SomeoneSerge — CHANGES_REQUESTED** [review]
Private keys seem to be a part of the persisted state, but I cannot see how their presence reflects in GUI, or how are they ever accessed for anything after the user has been set up.

- [ ] Is there still any need for storing the private key at all?
- [ ] When the private key is persisted, consider making this explicit for the user: maybe a red "Forget private key" button

File: templates/base.html (hunk 1 above)
Comment (ferrine) [reply templates/base.html:195]: What do you mean?

**Comment by ferrine** [comment]
> Private keys seem to be a part of the persisted state, but I cannot see how their presence reflects in GUI, or how are they ever accessed for anything after the user has been set up.
> 
> * [ ]  Is there still any need for storing the private key at all?
> * [ ]  When the private key is persisted, consider making this explicit for the user: maybe a red "Forget private key" button

* Private key is used to include in email config
* Not storing private key at all means we can't send client config (if he doesn't share public key first)
* Like the idea to add a button, but not as a part of this PR

File: templates/base.html (hunk 1 above)
Comment (ferrine) [reply templates/base.html]: Got ot

**Comment by SomeoneSerge** [comment]
> Like the idea to add a button, but not as a part of this PR

With this PR, the private key becomes `Maybe String` rather than `String`. This rather should reflect in the GUI

> Private key is used to include in email config

Oh

**Comment by ferrine** [comment]
@ngoduykhanh can you please have a look?

**Comment by ngoduykhanh** [comment]
@ferrine thank you for your contribution. The code changes look good to me but I have a few concerns/comments:
- What is the purpose of this feature? Is it for importing the existing WG config? If so, how do we manage the private key? We need it to generate a proper QR code/config file.
- I see you use `ParseKey` to validate the public keys and preshared keys. How about validating the duplication? We should not allow importing a public key that already exists. 
- [UI] Since the **Public Key** and **Preshared Key** are optional, I think we should adjust the form to hide them by default (using Collapse for example) because it is a bit long now.

**Comment by ferrine** [comment]
@ngoduykhanh 
The only feature I decided not to implement is 
> check QR code creation, raise an error if the private key is missing
It still generates the QR code but the config is incomplete (it was supposed to be so). Incomplete config is an expected behavior so I treat this as a feature.

**Comment by ferrine** [comment]
This is a simple way to tell there is no QR
image

**Comment by ferrine** [comment]
an updated version disables Scan button
image

File: templates/clients.html [hunk 2]
```diff
@@ -68,7 +68,7 @@ <h4 class="modal-title">QR Code</h4>
                     <span aria-hidden="true">&times;</span>
                 </button>
             </div>
-            <img id="qr_code" class="w-100" style="image-rendering: pixelated;" src="" alt="QR code" />
+            <img id="qr_code" class="w-100" style="image-rendering: pixelated;" src="" alt="QR code not available" />
```
Comment (ngoduykhanh) [inline templates/clients.html:71]: Can you revert this line? It will be always `QR code not available` even the QR code is available.

**Review by ngoduykhanh — CHANGES_REQUESTED** [review]
Thanks @ferrine for addressing the issues. I have one more comment below. Other changes look good.

**Comment by ferrine** [comment]
Done

**Review by ngoduykhanh — APPROVED** [review]

**Pull request merged**